csharp/ql/src/semmle/code/csharp/dataflow/internal/rangeanalysis/SignAnalysisSpecific.qll

```diff
@@ -0,0 +1,191 @@
+
+  float getNonIntegerValue(Expr e) {
+    result = e.(LongLiteral).getValue().toFloat() or
+    result = e.(RealLiteral).getValue().toFloat()
+  }
+
+  predicate containerSizeAccess(Expr e) {
+    propertyOverrides(e.(PropertyAccess).getTarget(), "System.Collections.Generic.IEnumerable<>",
+      "Count") or
+    propertyOverrides(e.(PropertyAccess).getTarget(), "System.Collections.ICollection", "Count") or
+    propertyOverrides(e.(PropertyAccess).getTarget(), "System.String", "Length") or
+    propertyOverrides(e.(PropertyAccess).getTarget(), "System.Array", "Length") or
+    e instanceof Linq::CountCall
+  }
+
+  class NumericOrCharType extends Type {
+    NumericOrCharType() {
+      this instanceof CharType or
+      this instanceof IntegralType or
+      this instanceof FloatingPointType or
+      this instanceof DecimalType
+    }
+  }
+
+  predicate unknownIntegerAccess(Expr e) {
+    // array access, indexer access
+    e instanceof ElementAccess and e.getType() instanceof NumericOrCharType
+    or
+    // property access
+    e instanceof PropertyAccess and e.getType() instanceof NumericOrCharType
+    or
+    //method call, local function call, ctor call, ...
+    e instanceof Call and e.getType() instanceof NumericOrCharType
+  }
+
+  Sign explicitSsaDefSign(SsaVariable v) {
+    exists(AssignableDefinition def | def = v.(ExplicitDefinition).getADefinition() |
+      result = exprSign(def.getSource())
+      or
+      not exists(def.getSource()) and
+      not def.getElement() instanceof MutatorOperation
+      or
+      result = exprSign(def.getElement().(PostIncrExpr).getOperand()).inc()
+      or
+      result = exprSign(def.getElement().(PreIncrExpr).getOperand()).inc()
+      or
+      result = exprSign(def.getElement().(PostDecrExpr).getOperand()).dec()
+      or
+      result = exprSign(def.getElement().(PreDecrExpr).getOperand()).dec()
+    )
+  }
+
+  Sign implicitSsaDefSign(SsaVariable v) {
+    v =
+      any(ImplicitDefinition id |
+        result = fieldSign(id.getSourceVariable().getAssignable()) or
+        not id.getSourceVariable().getAssignable() instanceof Field
+      )
+  }
+
+  /** Gets a possible sign for `f`. */
+  Sign fieldSign(Field f) {
+    result = exprSign(f.getAnAssignedValue())
+    or
+    exists(PostIncrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).inc())
+    or
+    exists(PreIncrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).inc())
+    or
+    exists(PostDecrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).dec())
+    or
+    exists(PreDecrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).dec())
+    or
+    exists(AssignOperation a | a.getLValue() = f.getAnAccess() | result = exprSign(a))
+    or
+    f.fromSource() and not exists(f.getInitializer()) and result = TZero()
+  }
+
+  Sign specificSubExprSign(Expr e) {
+    result = exprSign(e.(AssignExpr).getRValue())
+    or
```

Comment:
If you add
```ql
or
result = exprSign(e.(AssignOperation).getExpandedAssignment())
```
you can simplify below by changing all
```ql
(e instanceof AssignOPExpr or e instanceof OPExpr)
```
to just
```ql
e instanceof OPExpr
```
and you can simplify the predicates `binaryOpLhsSign` and `binaryOpRhsSign` as well. This also means that code such as `var x = y += 1` will be analyzable.